python/lsst/daf/butler/core/quantum.py

```diff
@@ -205,70 +158,37 @@ def outputs(self) -> NamedKeyDict[DatasetType, List[DatasetRef]]:
-        must be present before a `Registry.addQuantum()` is called.
-
-        Parameters
-        ----------
-        ref : `DatasetRef`
-            Reference for a Dataset to add to the Quantum's predicted inputs.
-        """
-        self._predictedInputs.setdefault(ref.datasetType, []).append(ref)
-
-    def _markInputUsed(self, ref: DatasetRef) -> None:
-        """Mark an input as used.
-
-        This does not automatically update a `Registry`.
-        For that use `Registry.markInputUsed()` instead.
-        """
-        # First validate against predicted
-        if ref.datasetType not in self._predictedInputs:
-            raise ValueError(f"Dataset type {ref.datasetType.name} not in predicted inputs")
-        if ref not in self._predictedInputs[ref.datasetType]:
-            raise ValueError(f"Actual input {ref} was not predicted")
-        # Now insert as actual
-        self._actualInputs.setdefault(ref.datasetType, []).append(ref)
-
-    def addOutput(self, ref: DatasetRef) -> None:
-        """Add an output `DatasetRef` to the `Quantum`.
-
-        This does not automatically update a `Registry`; all `outputs`
-        must be present before a `Registry.addQuantum()` is called.
-
-        Parameters
-        ----------
-        ref : `DatasetRef`
-            Reference for a Dataset to add to the Quantum's outputs.
-        """
-        self._outputs.setdefault(ref.datasetType, []).append(ref)
-
-    @property
-    def id(self) -> Optional[int]:
-        """Unique (autoincrement) integer for this quantum (`int`).
-        """
-        return self._id
-
-    @property
-    def startTime(self) -> Optional[astropy.time.Time]:
-        """Begin timestamp for the execution of this quantum
-        (`astropy.time.Time`).
-        """
-        return self._startTime
-
-    @property
-    def endTime(self) -> Optional[astropy.time.Time]:
-        """End timestamp for the execution of this quantum
-        (`astropy.time.Time`).
-        """
-        return self._endTime
-
-    @property
-    def host(self) -> Optional[str]:
-        """Name of the system on which this quantum was executed (`str`).
-        """
-        return self._host
+        return copy.copy(self._outputs)
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, Quantum):
+            return False
+        for item in ("taskClass", "dataId", "initInputs", "predictedInputs", "outputs"):
+            if getattr(self, item) != getattr(other, item):
+                return False
+        return True
+
+    def __hash__(self) -> int:
+        cumHash = hash(self.taskClass)
+        cumHash ^= hash(self.dataId)
+        if isinstance(self.initInputs, abc.Mapping):
+            values = self.initInputs.values()
+        else:
+            values = self.initInputs
+        for item in values:
+            cumHash ^= hash(item)
+        if self.predictedInputs:
+            for item in self.predictedInputs.values():
```

Comment:
Are these always ordered identically across different instances?

Reply:
Good point. I switched this to calculating from a sorted list of hashes, which will always be the same. I also realized there was no need to do this on each hash call, as `Quantum` is not immutable, so I do the calculation once in the constructor.